lib/i18n/tasks/split_key.rb

```diff
@@ -5,68 +5,51 @@ module Tasks
     module SplitKey
       module_function
 
+      PARENTHESIS_PAIRS = %w({} [] () <>).freeze
+      START_KEYS = PARENTHESIS_PAIRS.to_set { |pair| pair[0] }.freeze
+      END_KEYS = PARENTHESIS_PAIRS.to_h { |pair| [pair[0], pair[1]] }.freeze
+      private_constant :PARENTHESIS_PAIRS, :START_KEYS, :END_KEYS
+
       # split a key by dots (.)
       # dots inside braces or parenthesis are not split on
       #
       # split_key 'a.b'      # => ['a', 'b']
       # split_key 'a.#{b.c}' # => ['a', '#{b.c}']
       # split_key 'a.b.c', 2 # => ['a', 'b.c']
       def split_key(key, max = Float::INFINITY)
-        parts = []
-        pos   = 0
         return [key] if max == 1
 
-        key_parts(key) do |part|
-          parts << part
-          pos += part.length + 1
-          if parts.length + 1 >= max
-            parts << key[pos..] unless pos == key.length
-            break
+        parts = []
```

Comment:
This does a lot more allocations than the previous version.
`split_key` is called a lot, what's the impact on  performance?

Reply:
It's actually faster on my machine (M1 Macbook Air):

On commit [0b4b483c82664f26c5696fb0f6aa1297356e4683](https://github.com/glebm/i18n-tasks/commit/0b4b483c82664f26c5696fb0f6aa1297356e4683) (latest main):
```
> Benchmark.measure { 1_000_000.times do ; I18n::Tasks::SplitKey.split_key('a.#{b.c}.d.<e.f>'); end }
#<Benchmark::Tms:0x0000000105396880 @cstime=0.0, @cutime=0.0, @label="", @real=4.406474999999773, @stime=0.005112000000000005, @total=4.400721000000001, @utime=4.395609>
```

On commit [61bc4fc4c25632798401599f6be6b7a45fef9945](https://github.com/glebm/i18n-tasks/commit/61bc4fc4c25632798401599f6be6b7a45fef9945) (this PR):
```
> Benchmark.measure { 1_000_000.times do ; I18n::Tasks::SplitKey.split_key('a.#{b.c}.d.<e.f>'); end }
#<Benchmark::Tms:0x0000000103042190 @cstime=0.0, @cutime=0.0, @label="", @real=3.340463999998974, @stime=0.005928000000000003, @total=3.336179, @utime=3.330251>
```

Reply:
Nice!